src/Orleans.Transactions/State/TransactionalState.cs

```diff
@@ -231,14 +240,23 @@ Task ITransactionalResource.Abort(long transactionId)
             if (this.metadata.StableVersion.TransactionId >= transactionId && this.metadata.HighestReadTransactionId >= wlb)
             {
                 // Logs already persisted, nothing to do here
+                if (this.logger.IsEnabled(LogLevel.Debug)) this.logger.Debug("PersistPrepare TransactionId {1} ignored. Already persisted", transactionId);
                 return true;
             }
 
             List<PendingTransactionState<TState>> pending = this.log.Select(kvp => new PendingTransactionState<TState>(kvp.Value.Version.ToString(), kvp.Key, kvp.Value.NewVal)).ToList();
+            Metadata metadata = new Metadata()
+            {
+                StableVersion = this.metadata.StableVersion,
+                HighestVersion = this.version,
+                HighestReadTransactionId = wlb
+            };
             this.metadata.HighestVersion = this.version;
             this.metadata.HighestReadTransactionId = wlb;
-            this.eTag = await this.storage.Persist(StateName, this.eTag, this.metadata.ToString(), pending);
+            this.eTag = await this.storage.Persist(StateName, this.eTag, metadata.ToString(), pending);
+            this.metadata = metadata;
 
```

Comment:
I'm a little confused by this section. Is the issue that `this.metadata` is mutated in the background somehow? Otherwise, why create a new instance of it if we're mutating both instances in the same manner.

The answer is probably obvious, but I'm missing something.

Reply:
nope, my mistake.  Intended to not change this.metadata unless the operation complete successfully.  Will fix.